// @cryptoscan/swap-sdk/dist/cjs/tests/createTransaction.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const createTransaction_js_1 = require("../utils/createTransaction.js");
(0, vitest_1.describe)('createTransaction', () => {
    (0, vitest_1.test)('should create a solana transfer transaction txn', async () => {
        const transaction = await (0, createTransaction_js_1.createTransaction)({
            type: 'transfer',
            network: "solana",
            coinAddress: 'So11111111111111111111111111111111111111112',
            from: "Hvt7wmdSEyDfptyojo4iiS3sroTPKAeKMAc4PNzXg7Ds",
            to: "DggKFxSoev2xJfMeb1CvgXZ8BsDUXb54bhFtaVMUT9nU",
            amount: 1,
        });
        (0, vitest_1.expect)(transaction).toBeDefined();
    }, 10_000);
    (0, vitest_1.test)('should create a solana transfer transaction txn', async () => {
        await new Promise(resolve => setTimeout(resolve, 2_000));
        const transaction = await (0, createTransaction_js_1.createTransaction)({
            type: 'transfer',
            network: "solana",
            coinAddress: 'So11111111111111111111111111111111111111112',
            from: "Hvt7wmdSEyDfptyojo4iiS3sroTPKAeKMAc4PNzXg7Ds",
            to: "DggKFxSoev2xJfMeb1CvgXZ8BsDUXb54bhFtaVMUT9nU",
            amount: 1,
        });
        (0, vitest_1.expect)(transaction).toBeDefined();
    }, 12_000);
    (0, vitest_1.test)('should create a solana swap transaction txn', async () => {
        await new Promise(resolve => setTimeout(resolve, 4_000));
        const transaction = await (0, createTransaction_js_1.createTransaction)({
            network: "solana",
            from: "So11111111111111111111111111111111111111112",
            to: "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v",
            amount: 1,
            walletAddress: '11111111111111111111111111111111',
        });
        (0, vitest_1.expect)(transaction).toBeDefined();
    }, 14_000);
    (0, vitest_1.test)('should create a pumpfun transaction txn', async () => {
        await new Promise(resolve => setTimeout(resolve, 6_000));
        const transaction = await (0, createTransaction_js_1.createTransaction)({
            network: "solana",
            service: "pumpfun",
            from: "So11111111111111111111111111111111111111112",
            to: "HJAoYbnsf16Z8ftk3SsuShKLQQgzmxAPu41RTpjjpump",
            amount: 1,
            walletAddress: '11111111111111111111111111111111',
        });
        (0, vitest_1.expect)(transaction).toBeDefined();
    }, 16_000);
});
